File: apps/studio/backend/app/models/copilot.py
"""Studio Copilot API models."""

from __future__ import annotations

from typing import Annotated, Any, Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field


class CopilotWsRequestPayload(BaseModel):
    """Incoming Copilot WebSocket request payload."""

    model_config = ConfigDict(extra="ignore")

    user_message: str
    # 会话身份契约(COPILOT_ASSIST-5):消息属于哪个前端会话标签。后端以
    # (skill, session) 隔离 SDK 对话;没有归属的消息在边界拒绝,绝不落进
    # "当前恰好活跃的对话"。
    session_id: str = Field(min_length=1)
    model_override: str | None = None
    # The selected copilot role (one per `role_kind=="copilot"` role in settings,
    # each auto-matched to its model group). Defaults to copilot_chat.
    role: str | None = None
    # Absolute imported/local workspace root. When present, the backend validates
    # it against Studio's skill index and uses it as the Copilot SDK cwd.
    workspace_root: str | None = None
    # Structured Golden-owned judge facts prepared before sending a Copilot Judge
    # chat turn. These are prompt context, not opaque refs hidden in prose.
    judge_context: dict[str, Any] | None = None


class CopilotEventBase(BaseModel):
    """Base model for streamed Copilot WebSocket events."""

    model_config = ConfigDict(extra="forbid")


class CopilotEventContextResolved(CopilotEventBase):
    """F4: echo, as the first streamed event, what context was injected this turn.

    Anti-hidden-prompt-magic (same spirit as F1 "不省略"): the UI shows a
    collapsible "injected context" card so the user can see exactly what the
    copilot was given before it starts.
    """

    type: Literal["context_resolved"] = "context_resolved"
    summary: str
    detail: str


class CopilotEventThinking(CopilotEventBase):
    """Streamed extended-thinking delta.

    F1 (copilot-assist mvp1-alignment): the whole reasoning trace is streamed —
    the UI may collapse it, but it must never be dropped. The frontend renders
    this as a collapsible "Thought" block.
    """

    type: Literal["thinking_delta"] = "thinking_delta"
    content: str


class CopilotEventText(CopilotEventBase):
    type: Literal["text_delta"] = "text_delta"
    content: str


class CopilotEventToolUseStart(CopilotEventBase):
    """F8: transcribes EVERY tool call the SDK reports, by its real name.

    Tool policy (what may run) is enforced at the SDK layer (allowed_tools /
    can_use_tool) — the transcript never editorializes, so `tool_name` is an
    open string, not the pre-allowed subset (the model legitimately runs
    read-only tools like Glob/Grep outside that list).
    """

    type: Literal["tool_use_start"] = "tool_use_start"
    # Open string: the SDK reports real tool names (incl. read-only tools and
    # studio MCP tools mcp__studio__<tool>); policy lives in SDK options.
    tool_name: str
    tool_input: dict[str, Any]


class CopilotEventToolUseResult(CopilotEventBase):
    type: Literal["tool_use_result"] = "tool_use_result"
    tool_name: str
    success: bool
    result_summary: str


class CopilotEventPatchProposed(CopilotEventBase):
    """F5 safe-write (model B): a copilot Write/Edit applied to a workspace file.

    Emitted before the SDK applies the edit (apply-then-review), carrying the
    pre-edit bytes so the UI can show an inline diff and a Reject can restore the
    original bytes through the Rust sole writer. ``after_content`` is best-effort
    for instant rendering; the authoritative applied content is the file on disk.
    """

    type: Literal["patch_proposed"] = "patch_proposed"
    tool_use_id: str
    tool_name: Literal["Write", "Edit"]
    path: str
    before_existed: bool
    before_content: str
    after_content: str
    before_hash: str | None
    after_hash: str
    diff: str
    checkpoint_id: str


class CopilotEventToolApprovalRequired(CopilotEventBase):
    """A copilot tool call held for human-in-the-loop approval (挂起式).

    Bash 一律审批;Read/Glob/Grep 仅在目标越出 workspace + 挂载目录时审批。
    ``can_use_tool`` awaits the user's verdict — approving lets the CLI run the
    tool itself so its result reaches the model's context. ``detail`` is the
    Bash command text or the out-of-fence path being read.
    """

    type: Literal["tool_approval_required"] = "tool_approval_required"
    tool_use_id: str
    tool_name: str
    detail: str


class CopilotEventToolApprovalTimedOut(CopilotEventBase):
    """A held tool call whose approval window expired before anyone answered.

    Carries ``tool_use_id`` because the point of the event is to name WHICH
    hold died. The timeout used to travel as a generic ``error``, which could
    say that something expired but not what — so the card it was about had no
    way to recognise itself and sat on "Waiting for approval." with live
    buttons, forever, while the task behind it had already been stopped
    (problem ledger CP7).

    The stop is not negotiable and not a denial: ``can_use_tool`` returns
    ``interrupt=True`` so the model halts, rather than handing it a refusal no
    human ever made and letting the turn run on (COPILOT_ASSIST-8).
    """

    type: Literal["tool_approval_timed_out"] = "tool_approval_timed_out"
    tool_use_id: str
    tool_name: str
    message: str


class CopilotEventDone(CopilotEventBase):
    type: Literal["done"] = "done"


class CopilotEventError(CopilotEventBase):
    type: Literal["error"] = "error"
    message: str
    error_code: str | None = None
    error_payload: dict[str, Any] | None = None


CopilotEvent: TypeAlias = Annotated[
    CopilotEventContextResolved
    | CopilotEventThinking
    | CopilotEventText
    | CopilotEventToolUseStart
    | CopilotEventToolUseResult
    | CopilotEventPatchProposed
    | CopilotEventToolApprovalRequired
    | CopilotEventToolApprovalTimedOut
    | CopilotEventDone
    | CopilotEventError,
    Field(discriminator="type"),
]


class CopilotToolApprovalRequest(BaseModel):
    """Approve or reject a held Copilot tool call."""

    model_config = ConfigDict(extra="forbid")

    tool_use_id: str
    approve: bool


class CopilotToolApprovalResponse(BaseModel):
    """Result of resolving a held Copilot tool call.

    ``resolved`` is False when the approval no longer exists (already resolved,
    timed out, or session reset). Execution happens in the CLI after approval,
    so there is no stdout/stderr here by design.
    """

    model_config = ConfigDict(extra="forbid")

    tool_use_id: str
    approved: bool
    resolved: bool
    message: str | None = None


class CopilotInterruptResponse(BaseModel):
    """Result of a copilot stop-button interrupt (R7-I).

    ``interrupted`` is False when there was no active turn to stop — the click
    landed after the turn already finished, which is a harmless no-op.
    """

    model_config = ConfigDict(extra="forbid")

    interrupted: bool


class CopilotSessionCloseRequest(BaseModel):
    """Close one frontend tab's backend SDK conversation (COPILOT_ASSIST-5)."""

    model_config = ConfigDict(extra="forbid")

    session_id: str = Field(min_length=1)


class CopilotSessionCloseResponse(BaseModel):
    """``closed`` counts dropped SDK clients — 0 when the tab never talked to
    the backend (a draft tab closed before its first message), which is a
    harmless no-op, not an error."""

    model_config = ConfigDict(extra="forbid")

    closed: int

File: apps/studio/frontend/src/types/copilot.ts
import type { JsonObject } from '../api/types'

export type CopilotEventStatus = 'pending' | 'running' | 'success' | 'error'

export interface CopilotEventBase {
  id: string
  status: CopilotEventStatus
  receivedAt: number
  raw: unknown
}

export interface CopilotTextDeltaEvent extends CopilotEventBase {
  type: 'text_delta'
  content: string
}

export interface CopilotToolUseStartEvent extends CopilotEventBase {
  type: 'tool_use_start'
  /** Real tool name as the SDK reports it — an open string: read-only tools
   * (Glob/Grep) and studio MCP tools (mcp__studio__<tool>) included. */
  tool_name: string
  tool_input: JsonObject
}

export interface CopilotToolUseResultEvent extends CopilotEventBase {
  type: 'tool_use_result'
  tool_name: string
  success: boolean
  result_summary: string
}

export interface CopilotDoneEvent extends CopilotEventBase {
  type: 'done'
}

export interface CopilotThinkingDeltaEvent extends CopilotEventBase {
  type: 'thinking_delta'
  content: string
}

export interface CopilotContextResolvedEvent extends CopilotEventBase {
  type: 'context_resolved'
  summary: string
  detail: string
}

export interface CopilotErrorEvent extends CopilotEventBase {
  type: 'error'
  message: string
}

/** F5 safe-write: a copilot Write/Edit applied; carries the diff + restore data. */
export interface CopilotPatchProposedEvent extends CopilotEventBase {
  type: 'patch_proposed'
  toolUseId: string
  toolName: 'Write' | 'Edit'
  path: string
  beforeExisted: boolean
  beforeContent: string
  afterContent: string
  beforeHash: string | null
  afterHash: string
  diff: string
  checkpointId: string
  /** Review state, driven locally by Accept/Reject. */
  review: 'pending' | 'accepted' | 'rejected'
}

/** A copilot tool call held for human approval (Bash / out-of-fence read).
 * Approving lets the CLI execute the tool itself; `detail` is the Bash
 * command text or the out-of-fence path being read. */
/**
 * How a held tool call ended.
 *
 * `timed_out` is here rather than in a field of its own because this one field
 * already answers "how did this hold end", and a second field answering the
 * same question is a second thing that can disagree with the first. Nobody
 * decided in that case — which is exactly why it is worth recording.
 */
export type ToolApprovalDecision = 'pending' | 'approved' | 'denied' | 'timed_out'

export interface CopilotToolApprovalRequiredEvent extends CopilotEventBase {
  type: 'tool_approval_required'
  toolUseId: string
  toolName: string
  detail: string
  /**
   * What the user decided, on the event itself — the session is written to disk
   * as JSON, so a decision kept anywhere else does not survive the panel being
   * collapsed, a session tab switch, or Restore chat (problem ledger CP6).
   */
  decision: ToolApprovalDecision
}

export interface CopilotUnknownEvent extends CopilotEventBase {
  type: 'unknown'
  payload: unknown
}

export type CopilotEvent =
  | CopilotTextDeltaEvent
  | CopilotThinkingDeltaEvent
  | CopilotContextResolvedEvent
  | CopilotToolUseStartEvent
  | CopilotToolUseResultEvent
  | CopilotPatchProposedEvent
  | CopilotToolApprovalRequiredEvent
  | CopilotDoneEvent
  | CopilotErrorEvent
  | CopilotUnknownEvent

export interface CopilotMessage {
  id: string
  role: 'user' | 'assistant' | 'system'
  content: string
  events: CopilotEvent[]
  status: CopilotEventStatus
  createdAt: number
}

/**
 * The `tool_use_id` of a hold the server just expired, or null for any other
 * record.
 *
 * This one stream record does not become an event of its own: it settles a
 * card that is already on screen. It is read here, beside the decoder, because
 * both answer the same question — what does this raw record mean — and reading
 * a wire shape in two places is how the two readings drift apart.
 */
export function readToolApprovalExpiry(raw: unknown): string | null {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    return null
  }
  const record = raw as Record<string, unknown>
  if (record.type !== 'tool_approval_timed_out' || typeof record.tool_use_id !== 'string') {
    return null
  }
  return record.tool_use_id || null
}

export function normalizeCopilotEvent(raw: unknown, id: string): CopilotEvent {
  const receivedAt = Date.now()
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    return { id, type: 'unknown', status: 'error', receivedAt, raw, payload: raw }
  }

  const record = raw as Record<string, unknown>
  if (
    record.type === 'context_resolved' &&
    typeof record.summary === 'string' &&
    typeof record.detail === 'string'
  ) {
    return {
      id,
      type: 'context_resolved',
      status: 'success',
      receivedAt,
      raw,
      summary: record.summary,
      detail: record.detail,
    }
  }
  if (record.type === 'thinking_delta' && typeof record.content === 'string') {
    return { id, type: 'thinking_delta', status: 'running', receivedAt, raw, content: record.content }
  }
  if (record.type === 'text_delta' && typeof record.content === 'string') {
    return { id, type: 'text_delta', status: 'running', receivedAt, raw, content: record.content }
  }
  if (record.type === 'tool_use_start' && typeof record.tool_name === 'string') {
    return {
      id,
      type: 'tool_use_start',
      status: 'running',
      receivedAt,
      raw,
      tool_name: record.tool_name,
      tool_input: isJsonObject(record.tool_input) ? record.tool_input : {},
    }
  }
  if (record.type === 'tool_use_result' && typeof record.tool_name === 'string') {
    return {
      id,
      type: 'tool_use_result',
      status: record.success === false ? 'error' : 'success',
      receivedAt,
      raw,
      tool_name: record.tool_name,
      success: record.success !== false,
      result_summary: typeof record.result_summary === 'string' ? record.result_summary : '',
    }
  }
  if (
    record.type === 'patch_proposed' &&
    typeof record.path === 'string' &&
    (record.tool_name === 'Write' || record.tool_name === 'Edit')
  ) {
    return {
      id,
      type: 'patch_proposed',
      status: 'success',
      receivedAt,
      raw,
      toolUseId: typeof record.tool_use_id === 'string' ? record.tool_use_id : '',
      toolName: record.tool_name,
      path: record.path,
      beforeExisted: record.before_existed === true,
      beforeContent: typeof record.before_content === 'string' ? record.before_content : '',
      afterContent: typeof record.after_content === 'string' ? record.after_content : '',
      beforeHash: typeof record.before_hash === 'string' ? record.before_hash : null,
      afterHash: typeof record.after_hash === 'string' ? record.after_hash : '',
      diff: typeof record.diff === 'string' ? record.diff : '',
      checkpointId: typeof record.checkpoint_id === 'string' ? record.checkpoint_id : '',
      review: 'pending',
    }
  }
  if (
    record.type === 'tool_approval_required' &&
    typeof record.tool_name === 'string' &&
    typeof record.detail === 'string'
  ) {
    return {
      id,
      type: 'tool_approval_required',
      status: 'pending',
      receivedAt,
      raw,
      toolUseId: typeof record.tool_use_id === 'string' ? record.tool_use_id : '',
      toolName: record.tool_name,
      detail: record.detail,
      // This function reads the LIVE stream only; a just-arrived approval is
      // pending by definition. Restoring a session does not come through here —
      // the persisted CopilotEvent objects are read back as they were written.
      decision: 'pending',
    }
  }
  if (record.type === 'done') {
    return { id, type: 'done', status: 'success', receivedAt, raw }
  }
  if (record.type === 'error' && typeof record.message === 'string') {
    return { id, type: 'error', status: 'error', receivedAt, raw, message: record.message }
  }
  return { id, type: 'unknown', status: 'error', receivedAt, raw, payload: raw }
}

function isJsonObject(value: unknown): value is JsonObject {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}
